cli/src/commands/split.rs

```diff
@@ -54,6 +54,31 @@ pub(crate) struct SplitArgs {
     /// The revision to split
     #[arg(long, short, default_value = "@")]
     revision: RevisionArg,
+    /// The revision to copy as the first part of the split (Experimental)
+    ///
+    /// This option's behavior may change in the future. Experience reports and
+    /// feedback are appreciated.
+    ///
+    /// With this option, the first part of the split will contain the changes
+    /// between the parent of the REVISION and the FROM revision. The second
+    /// part of the split will contain the changes between the FROM revision and
+    /// the REVISION revision.
+    ///
+    /// This is especially useful if the FROM revision is a past version of
+    /// REVISION, with its commit id obtained via `jj obslog` or `jj log
+    /// --at-operation`.
```

Comment:
Do we want a note saying that it probably doesn't do what you want if you pass it past version that has since been rebased?

I can see both diff and interdiff behaviors being useful, btw, but the interdiff behavior seems more useful, mostly because the typical use case is what the one described above.

For the regular diff/restore behavior (i.e. what's implemented), you would probably more likely do it based on a commit that's still visible, but then it's unclear why you would want to split the current commit to match a still-visible one.

Reply:
This is a good question.

In my mind, the main advantage of the regular restore behavior is that it's simpler to explain and understand, it doesn't create conflicts, and it works well in many cases. It is a good point thought that, when the difference does matter, the interdiff behavior might often be preferable (thought I'd guess it's not always).

If there was no cost to switching to interdiff, I'd do it (unless I can think of or remember some other reason not to). The main thing that is giving me pause is: what do we call it? How do we document it?

There is a simple approach to solve this that is not very elegant. It would be to keep this PR as is and add a `--interdiff --restore-from X` behavior. This would be easy to document, but it's a step away from "do what I mean" behavior.

Reply:
Sorry, I realized later that I think I described the cases poorly as "diff" vs "interdiff". I think a better classification is "snapshot" vs "diff". So what I previously called the "interdiff" mode could be something like `jj split --copy-diff-from=X` or something like that. I think that's a pretty natural way to think of it. What do you think?

Btw, it's implied that the diff is copied to the first commit of the split. I think we can just explain that without having to call the option `jj split --copy-first-commit-diff-from=X` or similar.

Reply:
Is it something that `jj rebase -r X --insert-before Y --reparent-descendants` would do?

`jj split --insert-diff-from`/`--insert-snapshot-from` might be better names. I find it's confusing that `jj split --restore-from` inserts a new arbitrary state in between.

Reply:
> Is it something that jj rebase -r X --insert-before Y --reparent-descendants would do?

In theory, it's similar, though (unless we figure out a better but still consistent behavior) it would create divergence if X was hidden and Y had the same change id.

In practice, I tried this (without `--reparent-descendants`) in the case of `X` being hidden and having children and it's quite weird and buggy. It seems to try to rebase the children of `X`. Moreover, it leads to

```console
$ jj op diff
thread 'main' panicked at /Users/ilyagr/dev/jj/lib/src/dag_walk.rs:116:13:
graph has cycle
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
From operation 16d658cd9cdf: push all branches to git remote origin
  To operation 1eb8c899850d: rebase commit c9ee0eb96f90b40e88c9a83fb3268bc16ab748f4
```


> jj split `--insert-diff-from`/`--insert-snapshot-from` might be better names.

At a glance, I like these quite a bit, thanks!

Reply:
> > Is it something that jj rebase -r X --insert-before Y --reparent-descendants would do?
> 
> In theory, it's similar, though (unless we figure out a better but still consistent behavior) it would create divergence if X was hidden and Y had the same change id.

Perhaps, I should use `jj duplicate` (in place of `rebase`.) 

fwiw, we should fix `op diff` bug separately. I guess it wouldn't be robust against diverged change ids.

Reply:
> Perhaps, I should use jj duplicate (in place of rebase.)

Indeed, that sounds great to me. `duplicate` makes sense for hidden commits just as well as non-hidden ones.

I'm beginning to think that `rebase` doesn't actually make much sense for hidden commits; we might want to forbid that outright and suggest `duplicate` instead. If a person really, really needs it (and wants to create divergence) they could un-hide the revisions before rebasing.

If `jj duplicate -r X --insert-before Y --reparent-descendants` worked, it would do exactly the same thing as `jj split --insert-snapshot-from`, I think. `jj duplicate -r X --insert-before Y` would correspond to `jj split --insert-diff-from`. I am not sure whether the `split` options would be useful if the `duplicate` worked for this usecase; their only real use would be didactic. Depending on when we implement the `duplicate` functionality, we might want them until then.

I am a bit confused about whether we want `duplicate -s` or `duplicate -b`, but that could be a completely separate discussion. We can make it work, but I think the result is confusing (I should experiment with it a bit).